hal/spi_api.h

```diff
@@ -19,220 +18,346 @@
+ * during a single call to spi_transfer().
+ *
+ * ### Continuous mode
+ * In this mode SS is kept asserted between words. For example with 4 bits words msb first,
+ * continuous mode would transmit 5 symbols: 0x9 0x4 0x8 0xF 0x0
+ * ```
+ *        ___   _   _   _   _   _   _   _   _   _   _   _   _   _   _   _   _   _   _   _   _
+ * MCLK      \_/ \_/ \_/ \_/ \_/ \_/ \_/ \_/ \_/ \_/ \_/ \_/ \_/ \_/ \_/ \_/ \_/ \_/ \_/ \_/
+ *        _                                                                                 _
+ * SS      \_______________________________________________________________________________/
+ *        _ ___         ___     ___         ___             _______________                 _
+ * DATA   _X   \_______/   \___/   \_______/   \___________/               \_______________X_
+ * _
+ * ```
+ * where non-continuous mode would require a "pause" of a defined duration between symbols thus
+ * reducing the overall throughput of the bus : 0x9 0x9 0x3 0x8
+ * ```
+ *        ___   _   _   _   _  ___   _   _   _   _  ___   _   _   _   _  ___   _   _   _   __
+ * MCLK      \_/ \_/ \_/ \_/ --   \_/ \_/ \_/ \_/ --   \_/ \_/ \_/ \_/ --   \_/ \_/ \_/ \_/
+ *        _                 _  _                 _  _                 _  _                 __
+ * SS      \_______________/ -- \_______________/ -- \_______________/ -- \_______________/
+ *        _ ___         _____  _____         _____  _         _________  _____             __
+ * DATA   _X   \_______/   X_--_X   \_______/   X_--_X_______/       X_--_X   \___________X__
+ * _
+ * ```
+ * ### Symbol/Word
+ * Even though documentations usually uses the lexeme `word` to mean the unit of data that is
+ * sent/received it is ambiguous as it may not correspond to an actual `word` on the platform.
+ * To avoid any confusion we will use in this API the lexeme `symbol` instead.
+ * ### SPI channel
+ * This is the virtual communication channel created between a master and its slave.
+ * @{
+ */
 
-#define SPI_EVENT_ERROR       (1 << 1)
-#define SPI_EVENT_COMPLETE    (1 << 2)
-#define SPI_EVENT_RX_OVERFLOW (1 << 3)
-#define SPI_EVENT_ALL         (SPI_EVENT_ERROR | SPI_EVENT_COMPLETE | SPI_EVENT_RX_OVERFLOW)
-
-#define SPI_EVENT_INTERNAL_TRANSFER_COMPLETE (1 << 30) // Internal flag to report that an event occurred
+/**
+ * This enumerates the possible transmission mode of a SPI peripheral.
+ *
+ * The SPI_MODE_QUAD_IO mode may also be backed by a QSPI peripheral in "legacy SPI" mode.
+ */
+typedef enum {
+    SPI_MODE_SIMPLEX, /**< Unidirectional communication on a single wire. */
+    SPI_MODE_HALF_DUPLEX, /**< Bidirectional communication on a single wire. */
+    SPI_MODE_FULL_DUPLEX, /**< Bidirectional communication on two wire (MISO/MOSI). */
+    SPI_MODE_DUAL_IO, /**< Half-duplex communication on 2 wires (transferring 2 bits per clock).*/
+    SPI_MODE_QUAD_IO, /**< Half-duplex communication on 4 wires (transferring 4 bits per clock).*/
```

Comment:
Which of these are required to be implemented? There should probably be a minimum capabilities section for spi_get_capabilities in defined behaviors.

Reply:
This is described in the `General requirements` section.
The capabilities structure only exposes booleans the optional modes.

Reply:
SPI_MODE_HALF_DUPLEX, SPI_MODE_QUAD_IO and SPI_MODE_DUAL_IO are out of the scope of this HAL specification. I would keep this simple and allow SPI_MODE_SIMPLEX to be used by just passing in NC for the MOSI or MISO pin.